### PR TITLE
Define soft-float PCS for AArch64

### PR DESCRIPTION
**Pull request by ostannard**
With the release of the Cortex-R82, we now have an AArch64 CPU which can be built without any floating-point instructions or registers, so I think we should re-consider our decision to not have a soft-float variant of the AArch64 PCS.

I've worded this as a variant over the base PCS, to keep it simple.

No compiler implements this yet, but clang does already allow passing floating-point types without an FPU, and seems to implement the first bullet point correctly. I'm working on LLVM patches which will implement bullet points 2 and 3.

GCC doesn't currently allow any use of floating-point types without an FPU for AArch64, so we don't need to worry about compatibility with existing code there.

For the fourth bullet point, I don't think we need to define the PCS for vector types, because these are only defined by the ACLE when the relevant hardware exists.

### CODE REVIEW

File: aapcs64/aapcs64.rst
Comment (smithp35) [inline aapcs64/aapcs64.rst:2038]: Does the implementation treat these as just structs of integers (of the appropriate size) for example a struct of two floats would be treated as if it were a struct of two integers.

From the specification
```
struct S {
  int x; // substitute for float
  int y; // substitute for float
};
```
Would be considered a homogenous aggregate, just not a homogenous floating point aggregate. I can't see anything in the argument passing rules that does anything special with this definition so it would behave as if a non-homogenous composite type. However for future changes would we prefer it to be defined as a homogenous aggregate or a non-homogenous composite?

If we do want to go down that route perhaps:
``Types that the base procedure call standard would consider HFAs or HVAs are treated as homogenous aggregates of signed Integral type with the same byte size and alignment."

File: aapcs64/aapcs64.rst
Comment (smithp35) [inline aapcs64/aapcs64.rst:2029]: I think this document and aapcs32 hasn't put variant in the title. Presumably inheriting it from "The standard variants". Perhaps just "Soft-float"

File: aapcs64/aapcs64.rst
Comment (smithp35) [inline aapcs64/aapcs64.rst:2032]: I think we could use stronger language here. Perhaps
The soft-float variant is defined for Arm v8-R implementations that do not have floating-point instructions or registers. This variant is incompatible with the base procedure call standard and toolchains are not required to support it.

**Review by smithp35 — COMMENTED** [review]
The variant definition looks good to me. I've made some comments/suggestions on the language, some of which could probably do with a second opinion so don't feel that you need to apply them. My intent for most of them is to try and use as similar language as the rest of the document.

File: aapcs64/aapcs64.rst
Comment (ostannard) [reply aapcs64/aapcs64.rst:2029]: Done

File: aapcs64/aapcs64.rst
Comment (ostannard) [reply aapcs64/aapcs64.rst:2032]: Done

File: aapcs64/aapcs64.rst
Comment (ostannard) [reply aapcs64/aapcs64.rst:2036]: Done, but I've dropped the word `signed` to avoid implying that the integer representation is meaningful, other then the bit-pattern matching the floating point value.

File: aapcs64/aapcs64.rst
Comment (ostannard) [reply aapcs64/aapcs64.rst:2038]: I agree that the original wording conflicts with the existing definition of `homogenous aggregate`, but I think that change could be confusing, since `homogenous aggregates of signed Integral type` isn't used anywhere else. Maybe it would be better to cut this down to just "No types are considered to be HFAs or HVAs"?

Comment (smithp35) [reply aapcs64/aapcs64.rst]: I'd be happy with just "No types are considered to be HFAs or HVAs" I think that is sufficient for someone to work out what they need to do by following the rest of the rules in the doc and avoids a potentially confusing definition.

**Review by smithp35 — APPROVED** [review]
I've approved the changes. I've added mmalcomson as a reviewer. Will wait to see if he has any further comments before merging.

**Review by mmalcomson — APPROVED** [review]
Appreciate waiting on me (especially given my holidays causing delays).
I don't really have anything to add -- from what I see we've had good discussion on trying to ensure it is clear the scope across which we expect this variant PCS to act.
I appreciate the work -- thanks both!

File: aapcs64/aapcs64.rst
Comment (sallyarmneale) [inline aapcs64/aapcs64.rst:2045]: Add the final comma in the list:

The calling convention for Short Vector, Scalable Vector, and Scalable Predicate machine types is left undefined.

**Review by sallyarmneale — COMMENTED** [review]
Minor changes required.

File: aapcs64/aapcs64.rst
Comment (ostannard) [reply aapcs64/aapcs64.rst:2045]: #277

**Pull request merged**
